src/actions.rs

```diff
@@ -0,0 +1,886 @@
+            PointerOrigin::WebElement(e) => WDActions::PointerOrigin::Element(
+                webdriver::common::WebElement(e.element_id().to_string()),
+            ),
+        }
+    }
+}
+
+/// The type of pointer.
+#[derive(Debug, Clone)]
+pub enum PointerActionType {
+    /// A mouse pointer device.
+    Mouse,
+    /// A pen device.
+    Pen,
+    /// A touch device.
+    Touch,
+}
+
+impl From<PointerActionType> for WDActions::PointerType {
+    fn from(pat: PointerActionType) -> Self {
+        match pat {
+            PointerActionType::Mouse => WDActions::PointerType::Mouse,
+            PointerActionType::Pen => WDActions::PointerType::Pen,
+            PointerActionType::Touch => WDActions::PointerType::Touch,
+        }
+    }
+}
+
+impl From<PointerAction> for WDActions::PointerActionItem {
+    fn from(pa: PointerAction) -> Self {
+        match pa {
+            PointerAction::Pause { duration } => WDActions::PointerActionItem::General(
+                WDActions::GeneralAction::Pause(WDActions::PauseAction {
+                    duration: Some(duration),
+                }),
+            ),
+            PointerAction::Down { button } => WDActions::PointerActionItem::Pointer(
+                WDActions::PointerAction::Down(WDActions::PointerDownAction { button }),
+            ),
+            PointerAction::Up { button } => WDActions::PointerActionItem::Pointer(
+                WDActions::PointerAction::Up(WDActions::PointerUpAction { button }),
+            ),
+            PointerAction::Move {
+                duration,
+                origin,
+                x,
+                y,
+            } => WDActions::PointerActionItem::Pointer(WDActions::PointerAction::Move(
+                WDActions::PointerMoveAction {
+                    duration: Some(duration),
+                    origin: origin.into(),
+                    x: Some(x),
+                    y: Some(y),
+                },
+            )),
+            PointerAction::Cancel => {
+                WDActions::PointerActionItem::Pointer(WDActions::PointerAction::Cancel)
+            }
+        }
+    }
+}
+
+/// A channel containing `Null` actions.
+#[derive(Debug, Clone)]
+pub struct NullActionChannel {
+    /// An identifier to distinguish this channel from others. Choose a meaningful string.
+    /// May be useful for debugging.
+    id: String,
+    /// The list of actions for this channel.
+    actions: Vec<NullAction>,
+}
+
+impl NullActionChannel {
+    /// Create a new NullActionChannel.
+    pub fn new(id: &str) -> Self {
+        Self {
+            id: id.to_string(),
+            actions: Vec::new(),
+        }
+    }
+
+    /// Add a pause action to this channel.
+    pub fn add_pause(&mut self, duration: Duration) {
+        self.add_action(NullAction::Pause {
+            duration: duration.as_millis() as u64,
+        });
+    }
+
+    /// Add the specified action to this channel.
+    pub fn add_action(&mut self, action: NullAction) {
```

Comment:
How about using the name `then` for this (these)?

Reply:
Also, let's return `&mut Self` from each of these so that they're easier to construct by chaining.